### DISCUSSION TITLE
Why commit typechain types?

### DISCUSSION THREAD

**Discussion by sebastiantf** [PaulRBerg/hardhat-template#96] [Q&A] [answered]
I see that you started checking in types generated by typechain into version control starting from this commit: https://github.com/paulrberg/solidity-template/commit/4eaa67ecca18f60415f82f98ac2d2fe7129a585c

Typechain [docs](https://github.com/dethcrypto/TypeChain#q-should-i-commit-generated-files-to-my-repository) suggests not to commit the generated types though.

Having the types checked in along with the contracts does seem good. But the biggest con of that I've noticed is having to keep track of all the types and factories in .gitignore like you've done in the [hifi packages](https://github.com/hifi-finance/hifi/blob/main/packages/protocol/.gitignore)

Would love to know what finally pushed you to this approach

**Comment by PaulRBerg** [comment, marked as answer]
Hey @sebastiantf, the rationale is that I want to nudge my template users to ship the types as part of the publicly distributed npm package. For instance, see the published [types](https://unpkg.com/browse/@hifi/amm@1.7.1/dist/types/) of the `@hifi/amm` package.

There are lots of benefits to having the types as part of the npm package - the main one being that your package users would enjoy type safety when working with your smart contracts.

That comes at the cost of having to somehow track the types over time, and committing them in version control is the best way to do that. I know it's a chore to add many additional lines in `.gitignore`, but there's a lot of third-party contracts that must be excluded from both git and npm, and I know no better way to achieve that than excluding the entire `types` directory by default and add bespoke rules to skip the types that should be kept.

**Reply by sebastiantf** [reply]
Thanks!

Maybe the .gitignore chore is something that can be handled by hardhat-packager too?

**Reply by PaulRBerg** [reply]
I don't think so, because `hardhat-packager` cannot know what other files you want to git ignore.

I know that this isn't ideal - I'm still investigating better options.

**Reply by sebastiantf** [reply]
Doesn't `hardhat-packager` have the `contracts` config? It can just append to the existing gitignore?

**Reply by PaulRBerg** [reply]
Hmm that's an interesting idea .. could you create a discussion for it in the `hardhat-packager` repo?

**Reply by sebastiantf** [reply]
Done: https://github.com/paulrberg/hardhat-packager/discussions/15
